### PR TITLE
[14.0][ADD] sale_coupon_sale_order_based

### PR DESCRIPTION
**Pull request by ivs-cetmix**
Sales Order based domain for Sales Coupon

### CODE REVIEW

**Comment by elvise** [comment]
@sbidoul can you check what happen here with the runboat? 

thanks in advance for your support

**Comment by sbidoul** [comment]
Runboat was enjoying a sunny week-end :) Nah, actually there was something super weird going on: https://github.com/OCA/oca-ci/issues/29, so I had to shut it down to recover.

I manually triggered a build for this PR.

**Comment by elvise** [comment]
@sbidoul thanks so much man! :)

**Review by francesco-ooops — APPROVED** [review]
Functional review ok!

**Review by elvise — APPROVED** [review]
LGTM!

**Comment by francesco-ooops** [comment]
@chienandalu any opinion on this one?

**Comment by francesco-ooops** [comment]
@OCA/crm-sales-marketing-maintainers

**Comment by elvise** [comment]
@OCA/crm-sales-marketing-maintainers  is there a good soul who can make a review? :)

**Comment by elvise** [comment]
@rousseldenis would be great if can check this PR :)

**Comment by elvise** [comment]
@pedrobaeza can you check this PR ? Thanks :)

**Review by chienandalu — APPROVED** [review]
Thanks! Tested and code looks good :+1:

**Comment by pedrobaeza** [comment]
For being consistent with other existing methods and for shorten the name, I would go with `sale_coupon_criteria_order_based`

**Comment by francesco-ooops** [comment]
alright, @isserver1 can you change?

File: sale_coupon_sale_order_based/tests/test_sale_coupon_sale_order_based.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
+from odoo.addons.sale_coupon.tests.common import TestSaleCouponCommon
+
+
+class TestSaleCouponSaleOrderBased(TestSaleCouponCommon):
+    def test_program_rules_order_based(self):
+        """
+        Check program rules based on order domain
+        """
+        partner_test = self.env["res.partner"].create(
```
Comment (simahawk) [inline sale_coupon_sale_order_based/tests/test_sale_coupon_sale_order_based.py]: pls set `tracking_disable=True` when creating tracked records

File: sale_coupon_sale_order_based/tests/test_sale_coupon_sale_order_based.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            {
+                "name": "Test",
+                "email": "test@example.com",
+            }
+        )
+        user_test = self.env["res.users"].create(
```
Comment (simahawk) [inline sale_coupon_sale_order_based/tests/test_sale_coupon_sale_order_based.py]: and set  `no_reset_password=True` too when creating users. BTW, do you really need another user?

File: sale_coupon_sale_order_based/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
+{
```
Comment (simahawk) [inline sale_coupon_sale_order_based/__manifest__.py]: license headers are missing everywhere ;)

File: sale_coupon_sale_order_based/__manifest__.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    "category": "Sale",
+    "website": "https://github.com/OCA/sale-promotion",
+    "depends": ["sale_coupon"],
+    "external_dependencies": {},
+    "data": ["views/coupon_program_view.xml"],
+    "demo": [],
+    "qweb": [],
+    "installable": True,
+    "auto_install": False,
+    "application": False,
```
Comment (simahawk) [inline sale_coupon_sale_order_based/__manifest__.py]:
```suggestion
```

File: sale_coupon_sale_order_based/tests/test_sale_coupon_sale_order_based.py (hunk 2 above)
Comment (GabbasovDinar) [reply sale_coupon_sale_order_based/tests/test_sale_coupon_sale_order_based.py]: new user for the purity of the test :)

**Comment by GabbasovDinar** [comment]
@simahawk done

File: sale_coupon_criteria_order_based/__manifest__.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
+# Copyright 2022 Ooops404
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+{
+    "name": "Sales Coupon based on Sales Order values",
+    "version": "14.0.1.0.0",
+    "summary": "Allows to set a sales order related domain for coupon",
+    "author": "Ooops404, Cetmix, Odoo Community Association (OCA)",
+    "contributors": "Ooops404, Cetmix",
+    "license": "LGPL-3",
```
Comment (simahawk) [inline sale_coupon_criteria_order_based/__manifest__.py]: header says AGPL, choose one :)

File: sale_coupon_criteria_order_based/models/coupon_program.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        help="Coupon program will work for the order with selected domain only",
+    )
+
+    def _is_valid_order(self, order):
+        """
+        Check that we can apply the coupon for current order
+        """
+        if self.rule_order_domain and self.rule_order_domain != "[]":
+            domain = ast.literal_eval(self.rule_order_domain) + [("id", "=", order.id)]
+            return bool(self.env["sale.order"].search_count(domain))
```
Comment (simahawk) [inline sale_coupon_criteria_order_based/models/coupon_program.py]: as you already have the order and you are filtering by id just apply the domain on it
```suggestion
            return bool(order.filtered_domain(domain))
```

File: sale_coupon_criteria_order_based/models/coupon_program.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        else:
+            return True
+
+    def _check_promo_code(self, order, coupon_code):
+        """
```
Comment (simahawk) [inline sale_coupon_criteria_order_based/models/coupon_program.py]: this sounds more like a comment, I'd not override the original docstring for this

File: sale_coupon_criteria_order_based/models/coupon_program.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        Override method to check coupon based on order
+        """
+        message = super()._check_promo_code(order, coupon_code)
+        if not self._is_valid_order(order):
+            message = {"error": _("The order doesn't have access to this reward.")}
+        return message
+
+    @api.model
+    def _filter_programs_from_common_rules(self, order, next_order=False):
+        """
```
Comment (simahawk) [inline sale_coupon_criteria_order_based/models/coupon_program.py]: same

File: sale_coupon_criteria_order_based/models/coupon_program.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        programs = super()._filter_programs_from_common_rules(
+            order, next_order=next_order
+        )
+
+        # Order requirement should not be checked if the coupon got generated by
+        # a promotion program
+        # (the requirement should have only be checked to generate the coupon)
+        if not next_order:
+            programs = programs and programs._filter_programs_on_order(order)
```
Comment (simahawk) [inline sale_coupon_criteria_order_based/models/coupon_program.py]:
```suggestion
        if not next_order and programs:
            programs = programs._filter_programs_on_order(order)
```

File: sale_coupon_criteria_order_based/readme/CONFIGURE.rst [hunk 10]
```diff
@@ -0,0 +1 @@
+* No configuration is required
```
Comment (simahawk) [inline sale_coupon_criteria_order_based/readme/CONFIGURE.rst]: you can trash this file

File: sale_coupon_criteria_order_based/readme/DESCRIPTION.rst [hunk 11]
```diff
@@ -0,0 +1,7 @@
+Sale Coupon domain based on Sale Order values
```
Comment (simahawk) [inline sale_coupon_criteria_order_based/readme/DESCRIPTION.rst]: the name comes from the manifest in generated readme

File: sale_coupon_criteria_order_based/readme/HISTORY.rst [hunk 12]
```diff
@@ -0,0 +1,4 @@
+14.0.1.0.0
```
Comment (simahawk) [inline sale_coupon_criteria_order_based/readme/HISTORY.rst]: trash this if you are not going to maintain it using towcrier

**Review by simahawk — CHANGES_REQUESTED** [review]
can you squash once your done w/ the changes?

**Comment by GabbasovDinar** [comment]
@simahawk ready for review

**Review by simahawk — APPROVED** [review]

**Pull request merged**